docs/design/proposals/references/README.md

```diff
@@ -0,0 +1,243 @@
+    // controller and the `External` field below should be non-nil.
+    KubernetesReference *ackcorev1alpha1.AWSResourceReferencce `json:"kubernetesReference,omitempty"
+
+    // External indicates the AWS external identifiers for the IAMRole. 
+    // This field can be an Name, ARN or id depending on API input shape and 
+    // is used for resources which are not managed by ACK controller.
+    // If nil, the `KubernetesReferences` field must not be nil.
+    External *ackv1alpha1.AWSIdentifiers `json:"external,omitempty"`
+}
+```
+
+External field is beneficial for following reasons:
+
+* Service controller does not exist for this type of resource or is planned in future. This leaves a way for the Go types to be future proof.
+* User wants to use a resource which has been created by a different infrastructure team(using another Kubernetes cluster /CFN/Terraform/console).
+
+#### Approach 2
+Leave the default field for the external case and have an additional field to specify Kubernetes reference.
+
+```
+type EndpointSpec {
+    EndpointName *string `json:"endpointName,omitempty"`,
+    // Use to specify AWS external identifier
+    IAMRole *string `json:"IAMRole,omitempty"`,
+    // Use to specify KubernetesReference
+    IAMRoleRef *ackv1alpha1.AWSResourceReference `json:"IAMRoleRef,omitempty"`,
+    ...
+}
+```
+Either IAMRole or IAMRoleRef needs to be non-nil. If both are specified by mistake, controller will give preference to external i.e. IAMRole. 
+
+Note: The default IAMRole field will be modified to be optional if it is required in AWS API.
+
+### Modifications to Generator Config 
+
+Introduce a new field “references” to generator config under resources which contain instructions to the code-generator about related resources. References would be a map keyed by path to field name which should be converted to a reference and value being a referenceConfig which consists instructions about the value to be substituted. ReferenceConfig will consist of the following fields:
+
+* value: full path to the value from the referenced resource object to be substituted for parent resource field name.
+    * Full path is needed because value might be a generated field like ARN, id (like vpc-id) which will be present in the Status or can be names (like modelName) which are provided by user as part of spec
+* APIVersion and Kind:
+    * Kind is pre-determined based on the AWS API parameter. So we dont need this to be input by the user
+    * APIVersion needs to be fixed at code generation time so that controller is not affected by updates to another service controller. Kubernetes follows a [hub and spoke design](https://book.kubebuilder.io/multiversion-tutorial/conversion-concepts.html#what-does-that-have-to-do-with-webhooks) allows converting CRs from one version to other using webhooks.
+
+```
+# Initial draft, expect to be changed
+
+resources:
+  Endpoint:
+    references:
```

Comment:
I am wondering whether we can find a way to get this information from the API spec, i.e. the model JSON, somehow. I know our friends at Azure Service Operator are trying to convince their teams to share some kind of metadata, like `vpcId points to VPC`, in the OpenAPI schema they publish.

I can imagine that model JSON is published for many SDKs to use is pretty constrained due to having to be very generic, but from service owner perspective, it might be more manageable for other tooling they need to maintain to have the reference metadata in the source of truth. Though it may not be worth it either since the change there is probably higher cost than here.

Reply:
@muvaf yeah, I've considered this. If all the AWS API models were consistent, this would be possible... :) However, the reality is that the APIs have inconsistent patterns in how they refer to other objects, which makes this kind of thing pretty difficult to standardize on.